### PR TITLE
New internal query result payload format: add protobuf payload definition

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR adds the definition of the new protobuf payload format for query results.

It also removes the need to manually wire up dependencies between `.proto` files and their corresponding generated `.pb.go` files in `Makefile`.

#### Which issue(s) this PR fixes or relates to

#4104 

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: Makefile
Comment (charleskorn) [inline Makefile:220]: This is better than it was before, but it's still not quite right: we won't regenerate `.pb.go` files if a transitive dependency `.proto` has changed.

For example, if `pkg/mimirpb/mimir.proto` changes, we will regenerate `pkg/mimirpb/mimir.pb.go`, but not `pkg/distributor/distributorpb/distributor.pb.go`, even though `pkg/distributor/distributorpb/distributor.proto` imports `mimirpb.proto`. Under most circumstances, this won't be an issue (either there's nothing that needs to change in `distributor.pb.go`, or the Go compiler will catch the problem). 

The safest thing to do here would be to depend on all `.proto` files, but that also feels like overkill. Thoughts?

File: pkg/querypb/query_response.proto
Comment (charleskorn) [inline pkg/querypb/query_response.proto:17]: This enum (`Status`) and `ErrorType` below need to stay in sync with the possible values defined in Prometheus.

Alternatively, if we don't want to maintain these here, we could send strings for `status` and `error_type`, but this would increase payload size and allocations (we'd allocate an instance of `"success"` for every query payload decoded).

File: pkg/querypb/query_response_extra_test.go
Comment (charleskorn) [inline pkg/querypb/query_response_extra_test.go:38]: I don't love what I've done here, very open to feedback and alternative suggestions.

**Review by pracucci — COMMENTED** [review]
Good job! Overall LGTM. I left few comments: in particular I would like to see the Histogram addressed before we merge it.

**Comment by pracucci** [comment]
Also be aware of this PR https://github.com/grafana/mimir/pull/4129 @krajorama just opened.

File: pkg/querypb/query_response.proto
Comment (krajorama) [reply pkg/querypb/query_response.proto:95]: agreed, you should be able to reuse the one in mimir.proto

File: pkg/querypb/query_response.proto
Comment (charleskorn) [reply pkg/querypb/query_response.proto]: Makes sense, addressed in 8d71462.

File: pkg/querypb/query_response.proto
Comment (charleskorn) [reply pkg/querypb/query_response.proto]: Makes sense, addressed in 8d71462.

**Review by jhesketh — APPROVED** [review]
looks like a good starting point to me :+1:

File: pkg/querypb/query_response.proto
Comment (charleskorn) [reply pkg/querypb/query_response.proto]: I've just started looking at implementing conversion from PromQL engine results to this new payload format, and it looks like PromQL will only ever return `histogram.FloatHistogram`s, never `histogram.Histogram`s (as [`promql.Point.H`](https://github.com/prometheus/prometheus/blob/1f0cc09579b7d04bbcb2b9746b7040da55791d48/promql/value.go#L110) is a `FloatHistogram`). Therefore a `QueryResponse` will only ever contain float histograms. 

However, this definition of `Histogram`, which is used for write requests, caters to both float and integer histograms. While we can just ignore all of the integer histogram-specific fields, this could could easily lead to mistakes. (The `oneof` fields are also awkward to work with and wouldn't be needed in a specialised types.) How would you feel about introducing a `FloatHistogram` message that is used by `QueryResponse` @pracucci @krajorama?

**Comment by charleskorn** [comment]
@krajorama I'd like to mark `Histogram.negative_spans` and `Histogram.positive_spans` as non-nullable (ie. add a `[(gogoproto.nullable) = false]` annotation) to remove the unnecessary pointers in the generated Golang types. 

Is that going to be problematic? The corresponding fields on both `histogram.Histogram` and `histogram.FloatHistogram` don't accept pointers.

File: pkg/querypb/query_response.proto
Comment (charleskorn) [reply pkg/querypb/query_response.proto]: (I've added a `FloatHistogram` message in adea9e4.)

**Review by pracucci — APPROVED** [review]
LGTM!

**Comment by krajorama** [comment]
> @krajorama I'd like to mark `Histogram.negative_spans` and `Histogram.positive_spans`'s values as non-nullable (ie. add a `[(gogoproto.nullable) = false]` annotation) to remove the unnecessary pointers in the generated Golang types. I've done this in [4c14901](https://github.com/grafana/mimir/commit/4c1490136b9b6770d0ae68671de7c8da02995186).
> 
> Is that going to be problematic? The corresponding fields on both `histogram.Histogram` and `histogram.FloatHistogram` don't accept pointers.

Strictly speaking it's not possible to know. Currently prometheus remote write uses [this function](https://github.com/prometheus/prometheus/blob/c70d85baed260f6013afd18d6cd0ffcac4339861/storage/remote/codec.go#L603) to encode spans, which means that field will not be empty. However this depends on the remote writer sending data to Mimir.

On the other hand it seems unlikely you'd send something like [{1,2}, null, {3,4}] as Spans list, so I think it's ok. Worth a try making the change in Prometheus as well!

File: pkg/querypb/query_response.proto
Comment (krajorama) [reply pkg/querypb/query_response.proto]: In general I think tests should protect against such mistakes, also the conversion from/to mimirpb.Histogram and (prometheus)histogram.Histogram and histogram.FloatHistogram will be localized into function , e.g. https://github.com/grafana/mimir/blob/3e80f787833ead5ac0a835fc94219ce9f723f0e0/pkg/mimirpb/compat.go#L241  

So I feel it's a bit too redundant and introduces extra code and things to keep in sync. I recommend not doing it.

Comment (charleskorn) [reply pkg/querypb/query_response.proto]: I'll revert it for now, but I'm also keen to benchmark it and see if a specialised type has any performance impact. (This will be for a later PR though.)

Comment (charleskorn) [reply pkg/querypb/query_response.proto]: I've [benchmarked](https://github.com/grafana/mimir/commit/3efffb0b62c0c23d430ae6d8c92e769518209ad9) encoding with the generic `Histogram` type and with a dedicated `FloatHistogram` type, and I think the results justify the extra effort:

```
name                                                                              old time/op    new time/op    delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                      1.03µs ± 0%    0.90µs ± 0%  -12.34%  (p=0.016 n=4+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                      535ns ± 0%     482ns ± 0%   -9.84%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10      862ns ± 0%     757ns ± 0%  -12.11%  (p=0.016 n=4+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                      493ns ± 0%     427ns ± 0%  -13.40%  (p=0.008 n=5+5)

name                                                                              old alloc/op   new alloc/op   delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                      1.22kB ± 0%    1.06kB ± 0%  -13.07%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                       624B ± 0%      544B ± 0%  -12.82%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10     1.10kB ± 0%    0.94kB ± 0%  -14.49%  (p=0.008 n=5+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                       536B ± 0%      456B ± 0%  -14.93%  (p=0.008 n=5+5)

name                                                                              old allocs/op  new allocs/op  delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                        17.0 ± 0%      13.0 ± 0%  -23.53%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                       10.0 ± 0%       8.0 ± 0%  -20.00%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10       15.0 ± 0%      11.0 ± 0%  -26.67%  (p=0.008 n=5+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                       9.00 ± 0%      7.00 ± 0%  -22.22%  (p=0.008 n=5+5)
```

`oneof` requires the use of a pointer value, so with the generic `Histogram` type, we have to perform two allocations per histogram while encoding. With the specialised `FloatHistogram` type, we avoid these allocations and their CPU and heap impact.

**Comment by charleskorn** [comment]
> Worth a try making the change in Prometheus as well!

Done: https://github.com/prometheus/prometheus/pull/11935

**Comment by pracucci** [comment]
> > @krajorama I'd like to mark `Histogram.negative_spans` and `Histogram.positive_spans`'s values as non-nullable (ie. add a `[(gogoproto.nullable) = false]` annotation) to remove the unnecessary pointers in the generated Golang types. I've done this in [4c14901](https://github.com/grafana/mimir/commit/4c1490136b9b6770d0ae68671de7c8da02995186).
> > Is that going to be problematic? The corresponding fields on both `histogram.Histogram` and `histogram.FloatHistogram` don't accept pointers.
> 
> Strictly speaking it's not possible to know. Currently prometheus remote write uses [this function](https://github.com/prometheus/prometheus/blob/c70d85baed260f6013afd18d6cd0ffcac4339861/storage/remote/codec.go#L603) to encode spans, which means that field will not be empty. However this depends on the remote writer sending data to Mimir.
> 
> On the other hand it seems unlikely you'd send something like [{1,2}, null, {3,4}] as Spans list, so I think it's ok. Worth a try making the change in Prometheus as well!

Let me merge the PR as is. Nothing prevent us to change it later (given native histograms are not yet fully supported in Mimir). I don't have enough context on this, so I will let @krajorama jump on it. But given Krajo will start work in few hours, let me merge it and unblock you. Then based on what you decide, either you or Krajo can make a follow up PR to eventually change it.

File: pkg/querypb/query_response.proto
Comment (krajorama) [reply pkg/querypb/query_response.proto]: ack

**Pull request merged**
